crates/cli/src/commands/format.rs

```diff
@@ -0,0 +1,244 @@
+    let body_with_prefix = prefix_lines(input, &" ".repeat(2));
+    let escaped_body = body_with_prefix.replace("\"", "\\\"");
+    // body: |
+    //   escaped_body
+    format!("|\n{escaped_body}")
+}
+
+fn prefix_lines(input: &str, prefix: &str) -> String {
+    input
+        .lines()
+        .map(|line| {
+            if line.is_empty() {
+                line.to_owned()
+            } else {
+                format!("{prefix}{line}")
+            }
+        })
+        .collect::<Vec<_>>()
+        .join("\n")
+}
+
+fn apply_grit_rewrite(input: &str, pattern: &str) -> Result<String> {
+    let resolver = GritModuleResolver::new();
+    let rich_pattern = resolver.make_pattern(pattern, None)?;
+
+    let compiled = rich_pattern
+        .compile(&BTreeMap::new(), None, None, None)
+        .map(|cr| cr.problem)
+        .with_context(|| "could not compile pattern")?;
+
+    let rich_file = RichFile::new(String::new(), input.to_owned());
+    let runtime = ExecutionContext::default();
+    for result in compiled.execute_file(&rich_file, &runtime) {
+        if let MatchResult::Rewrite(rewrite) = result {
+            let content = rewrite
+                .rewritten
+                .content
+                .ok_or_else(|| anyhow!("rewritten content is empty"))?;
+            return Ok(content);
+        }
+    }
+    bail!("no rewrite result after applying grit pattern")
+}
+
+fn format_pattern_as_hunk_changes(pattern: &ResolvedGritDefinition) -> Result<HunkChange> {
+    let formatted_grit_code = format_grit_code(&pattern.body)?;
+    let body_range = pattern
+        .config
+        .range
+        .ok_or_else(|| anyhow!("pattern doesn't have config range"))?;
+    Ok(HunkChange {
+        starting_byte: body_range.start_byte as usize,
+        ending_byte: body_range.end_byte as usize,
+        new_content: formatted_grit_code,
+    })
+}
+
+/// format grit code using `biome`
+fn format_grit_code(source: &str) -> Result<String> {
+    let parsed = biome_grit_parser::parse_grit(source);
+    ensure!(
+        parsed.diagnostics().is_empty(),
+        "biome couldn't parse: {}",
+        parsed
+            .diagnostics()
+            .iter()
+            .map(|diag| diag.message.to_string())
+            .collect::<Vec<_>>()
+            .join("\n")
+    );
+
+    let options = GritFormatOptions::default();
+    let doc = biome_grit_formatter::format_node(options, &parsed.syntax())
+        .with_context(|| "biome couldn't format")?;
+    Ok(doc.print()?.into_code())
+}
+
+/// Represent a hunk of text that needs to be changed
+#[derive(Debug)]
+struct HunkChange {
+    starting_byte: usize,
+    ending_byte: usize,
+    new_content: String,
+}
+
+/// returns a new string that applies hunk changes
+fn apply_hunk_changes(input: &str, mut hunks: Vec<HunkChange>) -> String {
+    if hunks.is_empty() {
+        return input.to_string();
+    }
+    hunks.sort_by_key(|hunk| hunk.starting_byte);
```

Comment:
Since you're sorting it, just reverse the order (apply last to first) and use replace_range.

Reply:
Nice idea, Done